crates/red_knot_python_semantic/src/semantic_index/use_def.rs

```diff
@@ -1,153 +1,231 @@
-//! Build a map from each use of a symbol to the definitions visible from that use, and the
-//! type-narrowing constraints that apply to each definition.
+//! First, some terminology:
+//!
+//! * a "binding" gives a new value to a variable. This includes many different Python statements
+//!   (assignment statements of course, but also imports, `def` and `class` statements, `as`
+//!   clauses in `with` and `except` statements, match patterns, and others) and even one
+//!   expression kind (named expressions). It notably does not include annotated assignment
+//!   statements without a right-hand side value; these do not assign any new value to the
+//!   variable. We consider function parameters to be bindings as well, since (from the perspective
+//!   of the function's internal scope), a function parameter begins the scope bound to a value.
+//!
+//! * a "declaration" establishes an upper bound type for the values that a variable may be
+//!   permitted to take on. Annotated assignment statements (with or without an RHS value) are
+//!   declarations; annotated function parameters are also declarations. We consider `def` and
+//!   `class` statements to also be declarations, so as to prohibit accidentally shadowing them.
+//!
+//! Annotated assignments with a right-hand side, and annotated function parameters, are both
+//! bindings and declarations.
+//!
+//! We use [`Definition`] as the universal term (and Salsa tracked struct) encompassing both
+//! bindings and declarations. (This sacrifices a bit of type safety in exchange for improved
+//! performance via fewer Salsa tracked structs and queries, since most declarations -- typed
+//! parameters and annotated assignments with RHS -- are both bindings and declarations.)
+//!
+//! At any given use of a variable, we can ask about both its "declared type" and its "inferred
+//! type". These may be different, but the inferred type must always be assignable to the declared
+//! type; that is, the declared type is always wider, and the inferred type may be more precise.
```

Comment:
I actually aim to make this true without the added qualifier. When we hit an invalid assignment and emit a diagnostic, we infer the declared type as the inferred type for that binding, which keeps this true (and allows an explicit annotation to override incorrect type inference).

I think the one case in the next PR where this isn't true is when we see conflicting declarations. In that case in the next PR I currently just don't validate assignments at all, so the inferred type could be outside either of the conflicting declared types. I can mention that caveat here.

Reply:
After discussion today I plan to change the conflicting-declarations diagnostic so it still considers the union to be the declared type and validates assignments against that, while also issuing a conflicting-declarations diagnostic. So with that change I think it will again be fully true that inferred type is always assignable to declared type, even in ill-typed code.

Reply:
Feel free to merge the PR as is. Creating a new PR to change the behavior might also be easier to reason and helps document the decision.